docs/proposals/volume-provisioning.md

```diff
@@ -290,19 +234,68 @@ Implementation tasks associated with adding a new API group:
 6.  Add `kubectl` tool support
 7.  Add API group into API server
 
+### Provisioning configuration
+
+With the scheme outlined above, both `storageClass.ProvisioningParameters` and
+`claim.Spec.Selector` can specify parameters of requested volumes, such as
+availability zone, volume type (SSD/HDD) or IOPS. It's up to the provisioner to
+consider if these parameters can be satisfied and report error when they can't.
+
 ### Provisioner interface changes
 
 The `Provisioner` interface will be modified so that `Provision` accepts new
 parameters for storage class and claim:
 
 ```go
 type Provisioner interface {
-  Provision(claim api.PersistentVolumeClaim, storageClass storageapi.StorageClass) (api.PersistentVolume, error)
+  Provision(claim *api.PersistentVolumeClaim, storageClass *storageapi.StorageClass) (*api.PersistentVolume, error)
 }
 ```
 
 The existing provisioner implementations will be modified to implement this
-new interface and become sensitive to parameters of storage classes.
+new interface and become sensitive to parameters of storage classes and
+claim.Spec.Selector.
+
+`volume.Plugin` interface gets a new `GetLabeler` method, which returns
+`volume.Labeler` interface, which provides list of labels for a given PV. This
+is to be used in an admission controller to fill labels of PVs.
+
+```go
+type LabelerVolumePlugin interface {
+	VolumePlugin
+	NewLabeler(class *StorageClass, spec *Spec) (Labeler, error)
+}
+
+type Labeler interface {
+        GetLabels() (map[string]string, error)
+}
+```
+
+### PV Admission Controller Changes
+
+As the admission controller is implemented now, it does not use volume plugins
+to add labels to PVs and it has hardcoded support for GCE and AWS only, adding
+zone to all PVs.
+
+The admission controller would be changed to implement these steps:
+
+1. The admission controller will look at `pv.Spec.Class` of all created PVs. If
+   the class is empty, the controller finds a generic volume plugin resposible
+   for handling the PV and continues in step 4 (i.e. mimics current behaviour,
+   just through volume plugins)
+2. If Class attribute is found, it finds referenced StorageClass instance.
+3. It finds volume plugin specified by found `storageClass.ProvisionerType`.
+4. It calls `GetLabeler(pv.Spec)` on the plugin and `GetLabels` on the returned
+   interface to get list of labels to apply to the PV.
+5. It applies the labels to the PV, forcefuly overwriting any already existing
+   labels.
+
+That means that every volume plugin must implement `Labeler` interface and
+any out-of-tree provisioner **must** come with an admission controller that fills
+the labels used by that particular provisioner.
+
+*That does not sound like out-of-tree provisioner... Should they run their own
```

Comment:
If out of tree provisioners implement the interface, then it should be inline with in-tree provisioners. No need to run their own admission controller.